### PR TITLE
Support building as a library 

### PR DESCRIPTION
**Pull request by mb-gh**
Makefile now builds the app and the library
2 sub-Makefiles added, one for the app and the other for the library.

I added some new files for the lib and did not change any existing logic.
But I had to move few code blocks between files.

The result gives a clear separation between files:

NCurses display and main: `main.cpp cui.cpp cui.h Makefile MakeApp.mk`
Library related files: `libnethogs.cpp libnethogs.h MakeLib.mk`
Common logic: other files

### CODE REVIEW

**Comment by mb-gh** [comment]
continuous-integration build fails because it does not support or not configured for  c++11

**Comment by raboof** [comment]
What restrictions does c++11 place on which versions of gcc/clang etc we support?

File: process.h [hunk 1]
```diff
@@ -143,4 +143,13 @@ void refreshconninode ();
 
 void procclean ();
 
+void getkbps (Process * curproc, float * recvd, float * sent);
```
Comment (raboof) [inline process.h]: It's not entirely obvious to me whether this is a presentation issue or a library issue.. perhaps put them in their own file at least?

Comment (mb-gh) [inline process.h]: The problem with these getters...() functions is that they **modify** the Process structure and they are not simple getters !
the 3 conversion functions could be moved.
Let me see if I can't find a better way to move them.

**Comment by mb-gh** [comment]
Have a look at the update. I put the getters as members of Process. It should be more logical. tell me what you think.
Regarding c++11, there should be no particular restriction as long as a faily recent version of gcc is used. I use only some obvious c++11 features.
Can you tell me about that version of gcc/clang do you support? and how to test?

**Comment by raboof** [comment]
I currently haven't given it much thought, I think we should support "reasonably recent" compilers and test them (as far as possible) on travis.

Perhaps "support compilers that ship with the current Debian Stable" might be a good rule of thumb?

**Comment by mb-gh** [comment]
I build on Debian Stable with GCC so that should be it.
I will try to setup Travis on my side.
May be you can see if you can change Travis settings on your repo to upgrade GCC version or something....

**Comment by raboof** [comment]
Would `-std=c++0x` be sufficient? That's supported on more GCC versions (http://gcc.gnu.org/projects/cxx0x.html)

**Comment by mb-gh** [comment]
It works with my GCC.
The build param is in MakeLib.mk if you want to change it.

**Comment by raboof** [comment]
Travis uses g++ (Ubuntu/Linaro 4.6.3-1ubuntu5) for linux and LLVM version 6.0 (clang-600.0.54) (based on LLVM 3.5svn) for mac, it seems to make sense to support those configurations, right?

**Comment by mb-gh** [comment]
If it builds successfully that's fine by me.
If not, we have to change the travis config file to use a more recent version.
If 4.6.3 don't work, 4.7 should probably work fine.
FYI, my gcc version is `gcc (Debian 4.9.2-10) 4.9.2
`

**Comment by mb-gh** [comment]
> Official C++11 support started in g++ 4.7, though C++0x features were supported way earlier.

**Comment by raboof** [comment]
When using C++0x it builds on ubuntu precise, but still not on mac: https://travis-ci.org/raboof/nethogs/builds/113664853

**Comment by mb-gh** [comment]
That error could be fixed easily I think.
Give me a moment to fix it

**Comment by mb-gh** [comment]
It should be OK now :)

**Comment by mb-gh** [comment]
something particular about the library design is that it uses async callbacks... meaning that the callbacks don't happen in the same thread that does start/stop of the monitoring...

This makes the use easier, but may not obvious for people who don't know about multithreading.
 If you think this is bad, it could be changed to some kind of polling mechanism, the same way libpcap is designed...

By the way: the gui is developed [here](https://github.com/mb-gh/nethogs_gui). Still work in progress.

File: Makefile [hunk 2]
```diff
@@ -1,86 +1,31 @@
...
-	install -d -m 755 $(DESTDIR)$(sbin)
-	install -m 755 nethogs $(DESTDIR)$(sbin)
-	install -d -m 755 $(DESTDIR)$(man8)
-	install -m 644 nethogs.8 $(DESTDIR)$(man8)
-	@echo
-	@echo "Installed nethogs to $(DESTDIR)$(sbin)"
-	@echo
-	@echo "You might have to add this directory to your PATH and/or refresh your shells' path cache with a command like 'hash -r'."
-
+	$(MAKE) -f MakeApp.mk $@
```
Comment (raboof) [inline Makefile]: is 'make install' not needed for libraries?

File: MakeLib.mk [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+.PHONY: uninstall
+
+install: libnethogs
+	install -d -m 755 $(DESTDIR)$(sbin)
+	install -m 755 libnethogs $(DESTDIR)$(sbin)
+	@echo
+	@echo "Installed libnethogs to $(DESTDIR)$(sbin)"
+	@echo
+	@echo "You might have to add this directory to your PATH and/or refresh your shells' path cache with a command like 'hash -r'."
```
Comment (raboof) [inline MakeLib.mk]: This 'install' target doesn't quite make sense for libraries, though, right?

**Comment by raboof** [comment]
There seems to be a lot of duplication between MakeApp.mk and MakeLib.mk - could there be a better way?

File: MakeLib.mk (hunk 3 above)
Comment (mb-gh) [inline MakeLib.mk]: Yes you are right.
It needs to be changed to lib dir.

**Comment by raboof** [comment]
As for coding style (use of whitespace etc) things are starting to get a bit inconsistent. Is there a particular set of conventions you use?

Let's at least agree on conventions and make all new code follow those, 'fixing' existing code clearly is outside the scope of this PR.

**Comment by mb-gh** [comment]
Not sure what's the ideal way. the 2 files don't build the same set of files and use different build parameters.

**Comment by mb-gh** [comment]
There are many style variations : 
- http://www.stroustrup.com/bs_faq2.html
- https://google.github.io/styleguide/cppguide.html
may be others...
As long as there is a defined style... I can change my code to comply.

**Comment by nibroc** [comment]
@raboof Have you considered using something like [ClangFormat](http://clang.llvm.org/docs/ClangFormat.html)? It would allow you to trivially reformat the entire project into a consistent style, and from then on out a simple policy of "run ClangFormat before pull requesting" would ensure future consistency.

**Comment by mb-gh** [comment]
I will definitely have a look at this and use if possible.

**Comment by raboof** [comment]
@nibroc I'd definitely be in favour of using such a tool - I'm using scalariform for scala projects and that works out great. Let's do that in a different PR though :)

**Comment by mb-gh** [comment]
Yes. I found that my IDE (codelite) have a code formatter plugin.. it's very easy to reformat code.
The 'hard' part is to define the style :)

**Comment by mb-gh** [comment]
There should be probably be doxygen syled comments for library public header file.
Be we can it as last step before or after the merge.
The GUI project shows how to use it: https://github.com/mb-gh/gnethogs

**Comment by raboof** [comment]
(It looks like there's some merge failures)

File: process.cpp [hunk 4]
```diff
@@ -98,6 +115,94 @@ int Process::getLastPacket()
...
+	  	sum_recv = 0;
+	gettotal(&sum_recv, &sum_sent);
+	*recvd = tokb(sum_recv);
+	*sent = tokb(sum_sent);
+}
+
+void Process::gettotalb(float * recvd, float * sent)
+{
+	u_int32_t sum_sent = 0,
+	  	sum_recv = 0;
```
Comment (raboof) [inline process.cpp]: Indentation is a bit of a mess, but I guess we'll fix that with #44 :)

**Comment by raboof** [comment]
There seems to be code duplication between main.cpp and libnethogs.cpp now (the main loop), and between process.cpp and libnethogs.cpp (summing/cleaning connections).

Perhaps it would be good to clarify in the README that while nethogs is now available as a library, we don't intend the API to be stable at this point and recommend linking statically?

**Comment by mb-gh** [comment]
I am trying to fix the conflicts...
About the README and static linking... I don't know a lot about it so let's follow your advice. I have to add the static build commands to the makefile though. it's only dynamic build now.

**Comment by mb-gh** [comment]
It seems static linking may not be as easy as you think.
I tried static linking in my app and I have issues.

Another solution to the issue of API instability would be to use version numbers as part of the libname. this is the same way other libraries do I think.
Also, there are ways to change to API without changing the library number.. (adding functions and keeping the old ones for compatibility etc...) 

I will keep the static lib generation in the makefile though.

**Comment by mb-gh** [comment]
About the code duplication thing...
It cannot be helped as much as I know...
Because of the differences between an app and a lib....
Ideally, the main would USE the library and not do it's own thing... but I did not want to do that to avoid impacting the code.
**Would you like me to change the main and transform it into a 'client' of the lib, it would make the code of the main much smaller?**

**Comment by mb-gh** [comment]
It does not mean a lot of additional code will change... but it does mean that if there is a bug in the lib... it will impact the main app... while as of now, they are kind of independant.. thus the duplication of the code...
I personally would go with the factorisation, but it's up to you at the end if you want to go the additional step.

**Comment by raboof** [comment]
OK, I get your point about not wanting to touch the 'production' code too much yet while the new 'library' code is being battle-tested. I think I can concur it might be wise to keep them separate for now, and later when the library approach is battle-tested some more start sharing more code again and making the cli version a consumer of the library like the gui.

**Comment by raboof** [comment]
It seems we're doing some things that are OK for a standalone app but not really for a library, like 'extern' values, so let's make it clear that libnethogs is still in 'experimental' phase, but with that caveat I think we can start moving forward and merge.

**Comment by mb-gh** [comment]
I am good with that :)

**Comment by mb-gh** [comment]
Another approach would be to create a branch of nethogs for production bug-fixes only,
and then take the liberty to do the full rework on the main branch... 
It means the introduction of a alpha/beta version of next major upgrade with all the changes we want to do... while keeping a branch on the current version to maintain it production-class until the next one is production ready.

**Pull request merged**